### PR TITLE
Setting message receivers with env variable

### PR DESCRIPTION
**Pull request by trushton**
@kevgo @hugo @martinjaime @helenmedrano 
Instead of using the `service-messages.yml` to let exocom know which messages a service receives, those service-messages are now passed to exocom via an environment variable.

Tests are not passing currently as there are no service-messages to be passed to the exocom test. If anyone has input on this that would be great.

This goes along with the PR for exo-run.

### CODE REVIEW

File: src/client-registry/client-registry.ls [hunk 1]
```diff
@@ -41,6 +46,12 @@ class ClientRegistry
     @subscribers = {}
 
 
+  set-message-receivers: ->
```
Comment (kevgo) [inline src/client-registry/client-registry.ls]: I would inline this method. It doesn't do anything this is reusable or separately testable. Also, please read `process.env.SERVICE_MESSAGES` in the main file (cli.ls), and provide it to this class as a constructor argument. Only the cli file should interact with the environment of the application directly.

File: src/client-registry/client-registry.ls [hunk 2]
```diff
@@ -8,6 +9,10 @@ class ClientRegistry
 
   ->
 
+    # List of messages that are received by the applications services
+    @receives = []
```
Comment (kevgo) [inline src/client-registry/client-registry.ls]: You define `@receives` as an array here, but later you use it as a hash. Changing types like this prevents V8 from doing certain optimizations (converting high-level, dynamically typed code into low-level, statically typed one at runtime and compiling it using LLVM). So let's avoid those issues. Variables should never change types even in dynamically typed environments.

**Comment by trushton** [comment]
@kevgo The information was already being stored in the client registry, but the requested changes have been made, and tests are passing now.

File: features/register-service.feature [hunk 3]
```diff
@@ -21,19 +21,20 @@ Feature: Manage new instances of services
           internal-namespace: 'foo'
           host: 'localhost'
           port: 4001
+          receives: 'users.create'
```
Comment (kevgo) [inline features/register-service.feature]: Services should not send this. They just say "I am a users service". Exocom already knows what messages a service receives.

File: features/register-service.feature [hunk 4]
```diff
@@ -21,19 +21,20 @@ Feature: Manage new instances of services
...
         id: '123'
       """
     Then ExoCom now knows about these services:
-      | NAME          | TYPE  | INTERNAL NAMESPACE | HOST      | PORT |
-      | users-service | users | foo                | localhost | 4001 |
+      | NAME          | TYPE  | INTERNAL NAMESPACE | HOST      | PORT | RECEIVES     |
```
Comment (kevgo) [inline features/register-service.feature]: I think what messages a service receives is not relevant for this test, so let's remove this column.

File: src/cli.ls [hunk 5]
```diff
@@ -34,7 +35,7 @@ on-error = (err) ->
 
 
 run = ->
-  exocom = new ExoCom!
+  exocom = new ExoCom jsonic(process.env.SERVICE_MESSAGES or {})
```
Comment (kevgo) [inline src/cli.ls]: Shouldn't the `)` be after `SERVICE_MESSAGES`?

File: src/client-registry/client-registry.ls [hunk 6]
```diff
@@ -6,7 +6,12 @@ require! {
 
 class ClientRegistry
 
-  ->
+  (@service-messages) ->
+
+    # List of messages that are received by the applications services
+    @receives = {}
+    for service in @service-messages
```
Comment (kevgo) [inline src/client-registry/client-registry.ls]: Please extract this logic into a `set-service-messages` method that takes in the `service-messages` structure and populates the `@receives` attribute. The reasons are:
- constructors should not do any work. Here you are massaging data, which is work
- objects should be super easy to construct in tests. This makes testing easy, and making testing easy is part of test-driven architecture and a goal for us. Because this object requires `@service-messages` in the constructor, a number of tests have to pass in `{}` now, even though they don't really care about service messages and want to test something else. The rule should be that any attribute can be `null` in tests, and tests only provide data that is necessary for the operation under test.

File: src/exocom.ls [hunk 7]
```diff
@@ -13,8 +13,9 @@ debug = require('debug')('exocom')
 
 class ExoCom extends EventEmitter
 
-  ->
-    @client-registry    = new ClientRegistry
+  (@service-messages = {}) ->
```
Comment (kevgo) [inline src/exocom.ls]: ExoCom doesn't use `@service-messages`, so it shouldn't change its value or assign a default value. Do this in the class that actually uses this variable, and just pass it on here.

File: src/cli.ls (hunk 5 above)
Comment (trushton) [inline src/cli.ls]: The reason for the `)` being where it is is that the jsonic function throws an error in the case that `process.env.SERVICE_MESSAGES` is `null`, such as when running the tests. This is because `process.env.SERVICE_MESSAGES` is set as a flag of the exocom docker container's run command created in exo-run.

Comment (kevgo) [inline src/cli.ls]: Okay. Shouldn't you give it an empty string, or `'{}'` then? Jsonic parses strings.

Comment (trushton) [inline src/cli.ls]: The reason for the ')' being where it is is that jsonic will throw an error if `SERVICE_MESSAGES` is `null`, which is the case when running tests that are not using the `SERVICE_MESSAGES` environment variable, or if the service receives no messages.  I thought this would be a safe guard in case there comes a point in the future that we have a service that does not receive any messages, but simply broadcasts them out without caring for a response. (Kind of like a UDP connection to the other services)

Comment (trushton) [inline src/cli.ls]: Not only strings, it can also parse objects! But if you would prefer the empty string or '{}', we can do that.

File: features/support/cli-world.ls [hunk 8]
```diff
@@ -20,6 +20,7 @@ CliWorld = !->
     env =
       EXOCOM_HTTP_PORT : ports.http-port
       EXOCOM_ZMQ_PORT : ports.zmq-port
+      SERVICE_MESSAGES : "[{name:users-service, receives:users.create}]"
```
Comment (trushton) [inline features/support/cli-world.ls]: This is required by the `register-service.feature`. Should I create a new function in steps to pass this value in dynamically as we probably don't want it hardcoded like this?

Comment (hugobho) [inline features/support/cli-world.ls]: It wouldn't be too hard to move this logic to `steps.ls` right? We would just add another parameter to `@create-exocom-instance`. It would probably make sense for this to be defined within the step that it's used rather than in the global file. If anything just so it's more clear in the future what we are testing exactly.

File: src/exocom.ls [hunk 9]
```diff
@@ -13,8 +13,10 @@ debug = require('debug')('exocom')
 
 class ExoCom extends EventEmitter
 
-  ->
+  (@service-messages) ->
+
     @client-registry    = new ClientRegistry
+      ..set-service-messages @service-messages
```
Comment (hugobho) [inline src/exocom.ls]: What do you think about passing`@service-messages` into the constructor? It seems like that's information that would be needed right away, and plus we would never want to modify the messages routing table right?

File: src/client-registry/client-registry.ls [hunk 10]
```diff
@@ -48,6 +51,11 @@ class ClientRegistry
       @add-routing-config service
 
 
+  set-service-messages: (service-messages or {}) ->
+    for service in service-messages
```
Comment (hugobho) [inline src/client-registry/client-registry.ls]: I think `{}` is unnecessary here. We already do `process.env.SERVICE_MESSAGES or {}` in `cli.ls` right?

File: src/client-registry/client-registry.ls [hunk 11]
```diff
@@ -48,6 +51,11 @@ class ClientRegistry
       @add-routing-config service
 
 
+  set-service-messages: (service-messages or {}) ->
```
Comment (hugobho) [inline src/client-registry/client-registry.ls]: I think `{}` is unnecessary here. We already do `process.env.SERVICE_MESSAGES or {}` in `cli.ls` right?

File: src/cli.ls (hunk 5 above)
Comment (kevgo) [inline src/cli.ls]: I think it only parses strings, at least that's what the documentation says. Let's try to stick with that if possible.

File: src/client-registry/client-registry.ls (hunk 11 above)
Comment (kevgo) [inline src/client-registry/client-registry.ls]: This is the place that uses the data structure, so it should decide what default value to use. Let's remove it the default from cli.ls.

Comment (kevgo) [inline src/client-registry/client-registry.ls]: The syntax for default values for arguments is `=`, not `or`.

File: src/exocom-spec.ls [hunk 12]
```diff
@@ -6,7 +6,7 @@ require! {
 describe 'ExoCom', ->
 
   before-each ->
-    @exocom = new ExoCom
+    @exocom = new ExoCom {}
```
Comment (kevgo) [inline src/exocom-spec.ls]: I think such empty arguments are no longer needed now. You should be able to call `new ExoCom` here without any arguments.

File: src/client-registry/client-registry-spec.ls [hunk 13]
```diff
@@ -7,7 +7,7 @@ require! {
 describe 'ClientRegistry', ->
 
   before-each ->
-    @client-registry = new ClientRegistry
+    @client-registry = new ClientRegistry {}
```
Comment (kevgo) [inline src/client-registry/client-registry-spec.ls]: Call ClientRegistry without arguments

File: src/exocom.ls (hunk 9 above)
Comment (kevgo) [inline src/exocom.ls]: I was the one making the case to extract it. But that was on Friday and I was pretty worn out. Looking at this again, I agree with @hugobho that we can inject this via the constructor. My initial concerns were that this would require providing empty hashes for this in tests, but I think with better default values for missing arguments that concern goes away. Sorry about flip-flopping here @trushton. The good news is that this is an easy fix.

File: features/steps/steps.ls [hunk 14]
```diff
@@ -49,7 +50,8 @@ module.exports = ->
       ..base-port = 5000
       ..get-port N (@exocom-zmq-port) ~>
       ..get-port N (@exocom-http-port) ~>
-        @create-exocom-instance http-port: @exocom-http-port, zmq-port: @exocom-zmq-port, ~>
+        service-messages = '{}'
```
Comment (trushton) [inline features/steps/steps.ls]: For some reason, setting default values for parameters does not work in `cli-world.ls`, so this has to be set for two of the steps here.

**Comment by trushton** [comment]
@kevgo This should be good for another review

File: src/client-registry/client-registry.ls [hunk 15]
```diff
@@ -1,12 +1,18 @@
...
 
 
 class ClientRegistry
 
-  ->
+  (@service-messages = {}) ->
+
+    # List of messages that are received by the applications services
+    @receives = {}
+    for service in jsonic(@service-messages)
```
Comment (kevgo) [inline src/client-registry/client-registry.ls]: Use an object comprehension here:

```ls
@receives = {[service.name, service.receives] for service in jsonic(@service-messages)}

File: src/exocom-spec.ls [hunk 16]
```diff
@@ -6,7 +6,7 @@ require! {
 describe 'ExoCom', ->
 
   before-each ->
-    @exocom = new ExoCom
+    @exocom = new ExoCom!
```
Comment (kevgo) [inline src/exocom-spec.ls]: Undo this change

File: src/client-registry/client-registry.ls [hunk 17]
```diff
@@ -60,8 +60,8 @@ class ClientRegistry
       name: service.name
       type: service.type
       internal-namespace: service.internal-namespace
-    service.receives or= @receives[service.name]
-    for message in service.receives
+    @receives[service.name] or= service.receives or= {}
```
Comment (trushton) [inline src/client-registry/client-registry.ls]: Feels slightly hacky. Some of the tests still use `service.receives` when they set up services, but that will be phased out. the empty hash is for the case that a service receives no messages. This can be improved, but I thought that may be better left for the PR in which we will be switching from registering with `service.name` to `service.type`

Comment (trushton) [inline src/client-registry/client-registry.ls]: For example, the `broadcasting-messages.feature`. We set up the service landscape with a table including a `receives` column, which then calls `@set-service-landscape` which performs a post request to the exocom instance to `set-service-configuration` which in turn calls `add-routing-config` for each service. So at that point, there is data in the `service.receives` field, but not in the `@receives[service.name]` object, because that was created through the env variable, not the post request

**Comment by trushton** [comment]
@kevgo Latest comments all addressed. Some tests were updated to use the new service__messages environment variable, as before they had still been registering to exocom through an http request.

File: features/broadcasting-messages.feature [hunk 18]
```diff
@@ -9,10 +9,10 @@ Feature: Broadcasting messages
 
 
   Background:
-    Given an ExoCom instance configured for the service landscape:
-      | NAME  | TYPE  | INTERNAL NAMESPACE | HOST      | PORT | SENDS        | RECEIVES     |
-      | web   | web   | web                | localhost | 3001 | create-user  | created-user |
-      | users | users | users              | localhost | 3002 | created-user | create-user  |
+    Given an ExoCom instance with routing information "[{name: web, receives: [created-user]}, {name: users, receives: [create-user]}]" configured for the service landscape:
```
Comment (kevgo) [inline features/broadcasting-messages.feature]: Here is how this should look like in the end, when everything is implemented:

```cucumber
Given an ExoCom instance configured with the routes:
  """
  [
    { 
      name: "web",
      sends: [...],
      receives: [...]
    },
    {
      name: "users",
      sends: ...
    }
  ]
  """
And a "web" instance 
And a "users" instance
```

The steps `And a "..." instance` create the instance and make it register with ExoCom.

File: features/configuring-services.feature [hunk 19]
```diff
@@ -19,20 +19,12 @@ Feature: Configuring services
         internal-namespace: 'foo'
         host: 'localhost'
         port: 3001
```
Comment (kevgo) [inline features/configuring-services.feature]: I think this whole spec can go away now, since the endpoint to POST routing to it goes away?

File: features/steps/steps.ls [hunk 20]
```diff
@@ -17,29 +17,27 @@ module.exports = ->
...
 
 
-  @Given /^an ExoCom instance$/, (done) ->
+  @Given /^an ExoCom instance(?: with routing information "([^"]*)")?$/, (service-messages, done) ->
     port-reservation
       ..base-port = 5000
       ..get-port N (@exocom-zmq-port) ~>
       ..get-port N (@exocom-http-port) ~>
-        @create-exocom-instance zmq-port: @exocom-zmq-port, http-port: @exocom-http-port, done
+        @create-exocom-instance {ports: {zmq-port: @exocom-zmq-port, http-port: @exocom-http-port}, service-messages: service-messages or '{}'}, done
```
Comment (kevgo) [inline features/steps/steps.ls]: No need to group some arguments under `ports:` here. Just provide `zmq-port`, `http-port`, and `service-messages` as named parameters directly here.

File: src/client-registry/client-registry-spec.ls [hunk 21]
```diff
@@ -7,7 +7,7 @@ require! {
 describe 'ClientRegistry', ->
 
   before-each ->
-    @client-registry = new ClientRegistry
+    @client-registry = new ClientRegistry service-messages: '{}'
```
Comment (kevgo) [inline src/client-registry/client-registry-spec.ls]: Does assigning default values not work? The ClientRegistry constructor defines default values. Either remove the definition of default values from ClientRegistry, or remove the default values here.

File: src/client-registry/client-registry.ls [hunk 22]
```diff
@@ -1,12 +1,16 @@
...
+  'jsonic'
   'remove-value'
   'require-yaml'
 }
 
 
 class ClientRegistry
 
-  ->
+  ({@service-messages = {}}) ->
```
Comment (kevgo) [inline src/client-registry/client-registry.ls]: If this doesn't work, does this work?

```ls
({@service-messages} = {service-messages: '{}'}) ->

Comment (kevgo) [inline src/client-registry/client-registry.ls]: or this one:

```ls
({@service-messages = '{}'} = {}) ->

File: src/client-registry/client-registry.ls [hunk 23]
```diff
@@ -56,8 +60,8 @@ class ClientRegistry
       name: service.name
       type: service.type
       internal-namespace: service.internal-namespace
-    service.receives or= @_look-up-service-messages service.type
-    for message in service.receives
+    @receives[service.name] or= {}
+    for message in @receives[service.name]
```
Comment (kevgo) [inline src/client-registry/client-registry.ls]: You can combine these two lines thanks to LiveScript's expression-oriented nature:

```ls
for message in (@receives[service.name] or= {})

File: src/exocom-spec.ls [hunk 24]
```diff
@@ -6,7 +6,7 @@ require! {
 describe 'ExoCom', ->
 
   before-each ->
-    @exocom = new ExoCom
+    @exocom = new ExoCom service-messages: '{}'
```
Comment (kevgo) [inline src/exocom-spec.ls]: Undo this change and let ExoCom use its default values for `service-messages`

File: src/exocom.ls [hunk 25]
```diff
@@ -13,8 +13,9 @@ debug = require('debug')('exocom')
 
 class ExoCom extends EventEmitter
 
-  ->
-    @client-registry    = new ClientRegistry
+  ({@service-messages}) ->
+
```
Comment (kevgo) [inline src/exocom.ls]: remove this empty line

File: src/exocom.ls [hunk 26]
```diff
@@ -13,8 +13,9 @@ debug = require('debug')('exocom')
...
+    @client-registry    = new ClientRegistry service-messages: @service-messages
```
Comment (kevgo) [inline src/exocom.ls]:
```ls
@client-registry    = new ClientRegistry {@service-messages}

File: features/broadcasting-messages.feature (hunk 18 above)
Comment (trushton) [inline features/broadcasting-messages.feature]: @kevgo, all other comments have been addressed. Currently only the `receives` field is passed to ExoCom by environment variable. I do agree that the `sends` field should be included here; but as to ease the process and not have too many things worked on at once, @hugobho and I believe it may be better to get this in with just the `receives` field being passed through, then add in Hugo's changes for WebSockets, then return and add the `sends` field to the `SERVICE_MESSAGES` environment variable. Adding the `sends` field will include verifying that a service is authorized to send specific messages, which does not really fall under the scope of this PR.

Comment (kevgo) [inline features/broadcasting-messages.feature]: Is this a long-winded way of saying you are going to make this change later in a separate PR? If so, please create an issue with my comment in it, and link to it here so that we can close this thread here.

File: features/steps/steps.ls [hunk 27]
```diff
@@ -17,29 +17,27 @@ module.exports = ->
...
 
 
-  @Given /^an ExoCom instance$/, (done) ->
+  @Given /^an ExoCom instance(?: with routing information "([^"]*)")?$/, (service-messages, done) ->
     port-reservation
       ..base-port = 5000
       ..get-port N (@exocom-zmq-port) ~>
       ..get-port N (@exocom-http-port) ~>
-        @create-exocom-instance zmq-port: @exocom-zmq-port, http-port: @exocom-http-port, done
+        @create-exocom-instance {zmq-port: @exocom-zmq-port, http-port: @exocom-http-port, service-messages: service-messages or '{}'}, done
```
Comment (kevgo) [inline features/steps/steps.ls]: The default value should be defined when `create-exocom-instance` is declared, not each time it is used.

File: src/client-registry/client-registry.ls [hunk 28]
```diff
@@ -56,8 +60,7 @@ class ClientRegistry
       name: service.name
       type: service.type
       internal-namespace: service.internal-namespace
-    service.receives or= @_look-up-service-messages service.type
-    for message in service.receives
+    for message in (@receives[service.name] or= {})
```
Comment (kevgo) [inline src/client-registry/client-registry.ls]: Can `or=` be changed to `or` here? You don't use `@receives[service.name]` anywhere in this block, so why change it here?
